### PR TITLE
Deposit / Withdraw

### PR DESCRIPTION
**Pull request by kristiyanpetkov**
In my PR i made deposit and withdraw functions.  Current amount added in the user account page.

### CODE REVIEW

File: task5/src/main/java/com/clouway/core/DataValidator.java [hunk 1]
```diff
@@ -37,4 +37,15 @@ public boolean isValid(String username, String password, String email) {
     }
     return result;
   }
+
+  public boolean isValid(String amount) {
+    boolean result = false;
+    String amounPattern = "[0-9]{1,13}(\\.[0-9]*)?";
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/core/DataValidator.java]: I think that `2.565488756` will be valid. Make it validate amount with 2 decimal points as valid. all others are not.

File: task5/src/main/java/com/clouway/core/UserRepository.java [hunk 2]
```diff
@@ -12,4 +12,10 @@
   boolean authenticate(String email, String password);
 
   void deleteAll();
+
+  void makeDeposit(double amount, String email);
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/core/UserRepository.java]: `deposit` is enough. In this case there is nothing wrong. But it is always better if naming can be shorter. This is suggestion.

File: task5/src/main/test/com/clouway/adapter/http/BankOperationHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+      oneOf(validator).isValid(amount);
+      will(returnValue(true));
+
+      oneOf(userRepository).makeWithdraw(amountInDouble, email);
+      oneOf(response).sendRedirect("/useraccount?email=" + email);
+    }});
+    bankOperationHandler.doPost(request, response);
+  }
+}
```
Comment (panayotkulchev) [inline task5/src/main/test/com/clouway/adapter/http/BankOperationHandlerTest.java:200]: Not all scenarios are covered.
- invalid amounts
- withraw with not enough money

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+
+  public BankOperationHandler(UserRepository userRepository, Validator validator) {
+    this.userRepository = userRepository;
+    this.validator = validator;
+  }
+
+  protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+    String operation = request.getParameter("operation");
+    String email = request.getParameter("email");
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java]: This looks like security leak. Here we have to use something userrepository.getCurrentUser(sessionId); this is the common practise;

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    if (!validator.isValid(amount)) {
+      response.sendRedirect("/useraccount?email=" + email);
+      return;
+    }
+
+    Double amountAsDouble = Double.valueOf(amount);
+
+    if (operation.equalsIgnoreCase("Deposit")) {
+      userRepository.makeDeposit(amountAsDouble, email);
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java]: `userRepostitory` takes care of the user. `deposit()` and `withdraw()` has to be in `FundsRepository`

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      response.sendRedirect("/useraccount?email=" + email);
+      return;
+    }
+
+    userRepository.makeWithdraw(amountAsDouble, email);
+    response.sendRedirect("/useraccount?email=" + email);
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java]: Security leak what has to fixed. Described in previous comment

Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java]: Messages for success and for errors should be diplayed to the user.

File: task5/src/main/java/com/clouway/core/DataValidator.java (hunk 1 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/core/DataValidator.java]: RegEx fixed. Now only double amount with 2 decimal points are accepted.

File: task5/src/main/java/com/clouway/core/UserRepository.java (hunk 2 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/core/UserRepository.java]: renamed.

File: task5/src/main/test/com/clouway/adapter/http/BankOperationHandlerTest.java (hunk 3 above)
Comment (kristiyanpetkov) [inline task5/src/main/test/com/clouway/adapter/http/BankOperationHandlerTest.java:200]: Missing scenarios covered.

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java (hunk 4 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java]: Security leak fixed.

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java (hunk 5 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java]: `getBalance()`, `deposit()` and `withdraw()` moved to `FundsRepository`

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java (hunk 6 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java]: Security leak fixed. Messages for success added.

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    if (!validator.isValid(amount)) {
+      response.sendRedirect("/useraccount?message=Amount not valid. Example for a valid amount: 123.23 or 100.00");
+      return;
+    }
+
+    Double amountAsDouble = Double.valueOf(amount);
+
+    if (operation.equalsIgnoreCase("Deposit")) {
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java:55]: I would do deposit and withdraw in two separate pages with two controllers. Only suggestion

File: task5/src/main/java/com/clouway/adapter/http/UserAccount.java [hunk 8]
```diff
@@ -16,20 +20,43 @@
...
+    this.cookieFinder = cookieFinder;
+  }
 
   protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+    Cookie[] cookies = request.getCookies();
+    Cookie cookie = cookieFinder.find(cookies);
+    String sessionId = cookie.getValue();
+    String email = sessionRepository.getCurrentUserEmail(sessionId);
+    String message = request.getParameter("message");
     PrintWriter out = response.getWriter();
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/adapter/http/UserAccount.java:39]: Can put writer in the private method

File: task5/src/main/java/com/clouway/core/FundsRepository.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.clouway.core;
+
+/**
+ * Created by Kristiyan Petkov  <kristiqn.l.petkov@gmail.com> on 09.06.16.
+ */
+public interface FundsRepository {
+
+  void deposit(double amount, String email);
```
Comment (panayotkulchev) [inline task5/src/main/java/com/clouway/core/FundsRepository.java:8]: should be deposit(amount, userId); In this case email is the userId and it fits but usually is the id in DB. - Only for information

File: task5/src/main/test/com/clouway/adapter/http/BankOperationHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    bankOperationHandler = new BankOperationHandler(fundsRepository, validator, cookieFinder, sessionRepository);
+    amount = "200.00";
+    email = "admin@abv.bg";
+    amountInDouble = 200.00D;
+    cookies = new Cookie[]{};
+    cookie = new Cookie("sessionId", "3131-3344-6667-8843");
+  }
+
+  @Test
+  public void makeDepositHappyPath() throws Exception {
```
Comment (panayotkulchev) [inline task5/src/main/test/com/clouway/adapter/http/BankOperationHandlerTest.java]: depositMoney - is enought. we already discussed shorter names are better. suggestion

File: task5/src/main/test/com/clouway/adapter/persistence/PersistentFundsRepositoryTest.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  public void getBalance() throws Exception {
+    fundsRepository.deposit(290.00, email);
+
+    Double expected = 290.00;
+    Double actual = fundsRepository.getBalance(email);
+    assertEquals(expected, actual);
+  }
+
+  @Test
+  public void withdraw() throws Exception {
```
Comment (panayotkulchev) [inline task5/src/main/test/com/clouway/adapter/persistence/PersistentFundsRepositoryTest.java:45]: test case ware current amount is 0 and you withdraw money?

**Comment by panayotkulchev** [comment]
After fix comments LGTM!

File: task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java (hunk 7 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/adapter/http/BankOperationHandler.java:55]: Okay. I will have it in mind.

File: task5/src/main/java/com/clouway/adapter/http/UserAccount.java (hunk 8 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/adapter/http/UserAccount.java:39]: Done.

File: task5/src/main/java/com/clouway/core/FundsRepository.java (hunk 9 above)
Comment (kristiyanpetkov) [inline task5/src/main/java/com/clouway/core/FundsRepository.java:8]: Okay.

File: task5/src/main/test/com/clouway/adapter/persistence/PersistentFundsRepositoryTest.java (hunk 11 above)
Comment (kristiyanpetkov) [inline task5/src/main/test/com/clouway/adapter/persistence/PersistentFundsRepositoryTest.java:45]: Added.

**Pull request merged**